### PR TITLE
[HDF5] Add submodel parts

### PR DESCRIPTION
**Pull request by sunethwarna**
This PR adds submodel parts to xdmf output, so if the main model part is used as the model part for hdf5 output, then all the submodel parts can be read seperately inside paraview. No need of seperate processes with seperate model part outputs. This reduces the size of output since same information is used to create submodel parts for xdmf visualization in paraview.

But if someone wants to output element and condition data for submodel parts, then seperate hdf5 output processes need to be used, since hdf5 process only stores its condition and elemental data values, not that of submodel parts.

Multiblock inspector can be used to seperately visualize submodel parts, their elements, and their conditions.

This PR requires #6254.

There are duplicated commits from #6254, since this is built on top of that PR.

image

### CODE REVIEW

File: applications/HDF5Application/custom_processes/hdf5_xdmf_connectivities_writer_process.cpp [hunk 1]
```diff
@@ -52,9 +52,29 @@ void XdmfConnectivitiesWriterProcess::Execute()
...
 
+    KRATOS_ERROR_IF(mpFile->HasPath(mPrefix + "/Xdmf/SubModelParts")) << "Path \"" << mPrefix + "/Xdmf/SubModelParts\" exists." << std::endl;
+    CreateXdmfConnectivitiesForSubModelParts(mPrefix + "/SubModelParts", mPrefix + "/Xdmf/SubModelParts");
+
     KRATOS_CATCH("");
 }
 
+void XdmfConnectivitiesWriterProcess::CreateXdmfConnectivitiesForSubModelParts(const std::string& rPath, const std::string& rDestinationPrefix) const
+{
+    for (const auto& name : mpFile->GetGroupNames(rPath))
```
Comment (philbucher) [inline applications/HDF5Application/custom_processes/hdf5_xdmf_connectivities_writer_process.cpp]:
```suggestion
    for (const auto& r_name : mpFile->GetGroupNames(rPath))
```

File: applications/HDF5Application/custom_processes/hdf5_xdmf_connectivities_writer_process.cpp [hunk 2]
```diff
@@ -52,9 +52,29 @@ void XdmfConnectivitiesWriterProcess::Execute()
...
+    {
+        if (name == "Conditions" ||  name == "Elements")
+        {
+            for (const auto& item_name : mpFile->GetGroupNames(rPath + "/" + name))
```
Comment (philbucher) [inline applications/HDF5Application/custom_processes/hdf5_xdmf_connectivities_writer_process.cpp]:
```suggestion
            for (const auto& r_item_name : mpFile->GetGroupNames(rPath + "/" + name))
```

File: applications/HDF5Application/custom_io/hdf5_model_part_io.cpp [hunk 3]
```diff
@@ -262,13 +262,21 @@ std::vector<std::size_t> ModelPartIO::ReadContainerIds(std::string const& rPath)
     return ids;
 }
 
-void ModelPartIO::WriteSubModelParts(ModelPart const& rModelPart)
+void ModelPartIO::WriteSubModelParts(ModelPart const& rModelPart, const std::string& GroupName)
 {
-    mpFile->AddPath(mPrefix + "/SubModelParts");
+    mpFile->AddPath(mPrefix + GroupName);
     for (auto it = rModelPart.SubModelPartsBegin(); it != rModelPart.SubModelPartsEnd(); ++it)
```
Comment (philbucher) [inline applications/HDF5Application/custom_io/hdf5_model_part_io.cpp]: I don't think you need iterators here

File: applications/HDF5Application/custom_io/hdf5_model_part_io.cpp [hunk 4]
```diff
@@ -277,13 +285,13 @@ void ModelPartIO::WriteSubModelParts(ModelPart const& rModelPart)
         }
         if (GlobalNumberOfElements(*it) > 0)
         {
-            WriteContainerIds(*mpFile, sub_model_part_path + "/ElementIds", it->Elements(), info);
-            StoreWriteInfo(sub_model_part_path + "/ElementIds", info);
+            ModelPartIO current_model_part_io(mpFile, sub_model_part_path);
+            current_model_part_io.WriteElements(it->Elements());
```
Comment (philbucher) [inline applications/HDF5Application/custom_io/hdf5_model_part_io.cpp]: this directly works with an h5 file? I am impressed

**Review by philbucher — COMMENTED** [review]
How does this affect the file size?
Are more things now written to the h5 file?

**Comment by sunethwarna** [comment]
@philbucher, this PR or the #6254 will not increase solution step h5 file size. This PR will add connvectivity Ids for all the submodel parts' elements and conditions in the initial h5 file, which increases the file size. So for large simulations with lots of timesteps, this increase is negligible when you weigh it with the advantage of having all the submodel parts in one h5 file. 
In the other hand, this will reduce overall size since now you dont have to have different sets of processes for submodel parts which write redundant solution step values, as well as a redundant initial h5 file with model part connectivity information.

So i think, overall this should reduce storage requirement. (that was one of the objectives of this PR)

File: applications/HDF5Application/custom_io/hdf5_model_part_io.cpp (hunk 4 above)
Comment (sunethwarna) [reply applications/HDF5Application/custom_io/hdf5_model_part_io.cpp]: It works :)....

**Comment by sunethwarna** [comment]
Hi, @philbucher , now ci fails at DEM application? Does it have to do something with potential app as well??

**Comment by philbucher** [comment]
did you run the tests in the Structural?
I think they are segfaulting

**Comment by sunethwarna** [comment]
I am trying to investigate whats going on with the CI as well.

File: applications/HDF5Application/python_scripts/core/xdmf.py [hunk 5]
```diff
@@ -446,7 +452,7 @@ def create_xml_element(self):
     def add_attribute(self, attr):
         """Add an XDMF Attribute (results data set) to each child grid."""
         for grid in self.grids:
-            grid.add_attribute(attr)
+                grid.add_attribute(attr)
```
Comment (philbucher) [inline applications/HDF5Application/python_scripts/core/xdmf.py]: is this intentional?

**Review by philbucher — DISMISSED** [review]
one question

File: applications/HDF5Application/python_scripts/core/xdmf.py (hunk 5 above)
Comment (sunethwarna) [reply applications/HDF5Application/python_scripts/core/xdmf.py]: nope, i reverted it. Thanks

**Comment by sunethwarna** [comment]
@philbucher could you reapprove?

**Review by philbucher — APPROVED** [review]

**Pull request merged**